File: java/src/main/java/com/benchmark/Benchmark.java
package com.benchmark;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Benchmark {
    
    public static List<BenchmarkResult> runAllBenchmarks() {
        List<BenchmarkResult> results = new ArrayList<>();
        
        System.out.println("Running CPU-intensive benchmarks...");
        results.add(benchmarkPrimeNumbers());
        results.add(benchmarkMatrixMultiplication());
        results.add(benchmarkCryptographicHashing());
        results.add(benchmarkMathOperations());
        
        System.out.println("Running memory-intensive benchmarks...");
        results.add(benchmarkLargeArraySort());
        results.add(benchmarkMemoryAllocation());
        results.add(benchmarkStringConcatenation());
        
        return results;
    }
    
    private static BenchmarkResult benchmarkPrimeNumbers() {
        long start = System.nanoTime();
        
        int count = 0;
        int limit = 100000;
        for (int i = 2; i <= limit; i++) {
            if (isPrime(i)) {
                count++;
            }
        }
        
        long duration = System.nanoTime() - start;
        double durationSeconds = duration / 1e9;
        
        return new BenchmarkResult(
            "Prime Numbers (up to 100k)",
            duration,
            0, // Javaでは正確なメモリ測定が困難
            count,
            count / durationSeconds
        );
    }
    
    private static boolean isPrime(int n) {
        if (n < 2) return false;
        for (int i = 2; i <= Math.sqrt(n); i++) {
            if (n % i == 0) return false;
        }
        return true;
    }
    
    private static BenchmarkResult benchmarkMatrixMultiplication() {
        long start = System.nanoTime();
        
        int size = 500;
        Random random = new Random(42); // 再現可能性のためのシード
        double[][] a = new double[size][size];
        double[][] b = new double[size][size];
        double[][] c = new double[size][size];
        
        // 行列初期化
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                a[i][j] = random.nextDouble();
                b[i][j] = random.nextDouble();
            }
        }
        
        // 行列乗算
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                for (int k = 0; k < size; k++) {
                    c[i][j] += a[i][k] * b[k][j];
                }
            }
        }
        
        long duration = System.nanoTime() - start;
        double durationSeconds = duration / 1e9;
        long operations = (long) size * size * size;
        
        return new BenchmarkResult(
            "Matrix Multiplication (500x500)",
            duration,
            0,
            operations,
            operations / durationSeconds
        );
    }
    
    private static BenchmarkResult benchmarkCryptographicHashing() {
        long start = System.nanoTime();
        
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] data = new byte[1024];
            Random random = new Random(42);
            random.nextBytes(data);
            
            int iterations = 50000;
            for (int i = 0; i < iterations; i++) {
                digest.reset();
                digest.update(data);
                digest.digest();
            }
            
            long duration = System.nanoTime() - start;
            double durationSeconds = duration / 1e9;
            
            return new BenchmarkResult(
                "SHA256 Hashing (50k iterations)",
                duration,
                0,
                iterations,
                iterations / durationSeconds
            );
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not available", e);
        }
    }
    
    private static BenchmarkResult benchmarkMathOperations() {
        long start = System.nanoTime();
        
        int iterations = 10000000;
        double result = 0.0;
        for (int i = 0; i < iterations; i++) {
            double x = i;
            result += Math.sin(x) * Math.cos(x) * Math.sqrt(x + 1);
        }
        
        long duration = System.nanoTime() - start;
        double durationSeconds = duration / 1e9;
        
        // resultを使用して最適化を防ぐ
        if (Double.isNaN(result)) {
            System.out.println("Unexpected NaN result");
        }
        
        return new BenchmarkResult(
            "Math Operations (10M iterations)",
            duration,
            0,
            iterations,
            iterations / durationSeconds
        );
    }
    
    private static BenchmarkResult benchmarkLargeArraySort() {
        long start = System.nanoTime();
        
        int size = 1000000;
        Integer[] data = new Integer[size];
        Random random = new Random(42);
        for (int i = 0; i < size; i++) {
            data[i] = random.nextInt(1000001);
        }
        
        Arrays.sort(data);
        
        long duration = System.nanoTime() - start;
        double durationSeconds = duration / 1e9;
        
        return new BenchmarkResult(
            "Large Array Sort (1M elements)",
            duration,
            0,
            size,
            size / durationSeconds
        );
    }
    
    private static BenchmarkResult benchmarkMemoryAllocation() {
        long start = System.nanoTime();
        
        int allocations = 100000;
        List<int[]> arrays = new ArrayList<>(allocations);
        
        for (int i = 0; i < allocations; i++) {
            int[] data = new int[256]; // 1KB相当のデータ
            Arrays.fill(data, i % 256);
            arrays.add(data);
        }
        
        long duration = System.nanoTime() - start;
        double durationSeconds = duration / 1e9;
        
        return new BenchmarkResult(
            "Memory Allocation (100k x 1KB)",
            duration,
            0,
            allocations,
            allocations / durationSeconds
        );
    }
    
    private static BenchmarkResult benchmarkStringConcatenation() {
        long start = System.nanoTime();
        
        int iterations = 50000;
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < iterations; i++) {
            result.append("iteration_").append(i).append("_");
        }
        
        long duration = System.nanoTime() - start;
        double durationSeconds = duration / 1e9;
        
        return new BenchmarkResult(
            "String Concatenation (50k iterations)",
            duration,
            0,
            iterations,
            iterations / durationSeconds
        );
    }
}

File: kotlin/src/main/kotlin/com/benchmark/Benchmark.kt
package com.benchmark

import java.security.MessageDigest
import kotlin.math.*
import kotlin.random.Random

object Benchmark {
    
    fun runAllBenchmarks(): List<BenchmarkResult> {
        val results = mutableListOf<BenchmarkResult>()
        
        println("Running CPU-intensive benchmarks...")
        results.add(benchmarkPrimeNumbers())
        results.add(benchmarkMatrixMultiplication())
        results.add(benchmarkCryptographicHashing())
        results.add(benchmarkMathOperations())
        
        println("Running memory-intensive benchmarks...")
        results.add(benchmarkLargeArraySort())
        results.add(benchmarkMemoryAllocation())
        results.add(benchmarkStringConcatenation())
        
        return results
    }
    
    private fun benchmarkPrimeNumbers(): BenchmarkResult {
        val start = System.nanoTime()
        
        var count = 0
        val limit = 100000
        for (i in 2..limit) {
            if (isPrime(i)) {
                count++
            }
        }
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        
        return BenchmarkResult(
            test = "Prime Numbers (up to 100k)",
            durationNs = duration,
            memoryBytes = 0, // Kotlinでは正確なメモリ測定が困難
            operations = count.toLong(),
            opsPerSec = count / durationSeconds
        )
    }
    
    private fun isPrime(n: Int): Boolean {
        if (n < 2) return false
        for (i in 2..sqrt(n.toDouble()).toInt()) {
            if (n % i == 0) return false
        }
        return true
    }
    
    private fun benchmarkMatrixMultiplication(): BenchmarkResult {
        val start = System.nanoTime()
        
        val size = 500
        val random = Random(42) // 再現可能性のためのシード
        val a = Array(size) { DoubleArray(size) { random.nextDouble() } }
        val b = Array(size) { DoubleArray(size) { random.nextDouble() } }
        val c = Array(size) { DoubleArray(size) }
        
        // 行列乗算
        for (i in 0 until size) {
            for (j in 0 until size) {
                for (k in 0 until size) {
                    c[i][j] += a[i][k] * b[k][j]
                }
            }
        }
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        val operations = size.toLong() * size * size
        
        return BenchmarkResult(
            test = "Matrix Multiplication (500x500)",
            durationNs = duration,
            memoryBytes = 0,
            operations = operations,
            opsPerSec = operations / durationSeconds
        )
    }
    
    private fun benchmarkCryptographicHashing(): BenchmarkResult {
        val start = System.nanoTime()
        
        val digest = MessageDigest.getInstance("SHA-256")
        val data = ByteArray(1024)
        Random(42).nextBytes(data)
        
        val iterations = 50000
        repeat(iterations) {
            digest.reset()
            digest.update(data)
            digest.digest()
        }
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        
        return BenchmarkResult(
            test = "SHA256 Hashing (50k iterations)",
            durationNs = duration,
            memoryBytes = 0,
            operations = iterations.toLong(),
            opsPerSec = iterations / durationSeconds
        )
    }
    
    private fun benchmarkMathOperations(): BenchmarkResult {
        val start = System.nanoTime()
        
        val iterations = 10000000
        var result = 0.0
        for (i in 0 until iterations) {
            val x = i.toDouble()
            result += sin(x) * cos(x) * sqrt(x + 1)
        }
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        
        // resultを使用して最適化を防ぐ
        if (result.isNaN()) {
            println("Unexpected NaN result")
        }
        
        return BenchmarkResult(
            test = "Math Operations (10M iterations)",
            durationNs = duration,
            memoryBytes = 0,
            operations = iterations.toLong(),
            opsPerSec = iterations / durationSeconds
        )
    }
    
    private fun benchmarkLargeArraySort(): BenchmarkResult {
        val start = System.nanoTime()
        
        val size = 1000000
        val data = IntArray(size) { Random(42).nextInt(1000001) }
        
        data.sort()
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        
        return BenchmarkResult(
            test = "Large Array Sort (1M elements)",
            durationNs = duration,
            memoryBytes = 0,
            operations = size.toLong(),
            opsPerSec = size / durationSeconds
        )
    }
    
    private fun benchmarkMemoryAllocation(): BenchmarkResult {
        val start = System.nanoTime()
        
        val allocations = 100000
        val arrays = mutableListOf<IntArray>()
        
        repeat(allocations) { i ->
            val data = IntArray(256) { i % 256 } // 1KB相当のデータ
            arrays.add(data)
        }
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        
        return BenchmarkResult(
            test = "Memory Allocation (100k x 1KB)",
            durationNs = duration,
            memoryBytes = 0,
            operations = allocations.toLong(),
            opsPerSec = allocations / durationSeconds
        )
    }
    
    private fun benchmarkStringConcatenation(): BenchmarkResult {
        val start = System.nanoTime()
        
        val iterations = 50000
        val result = StringBuilder()
        repeat(iterations) { i ->
            result.append("iteration_${i}_")
        }
        
        val duration = System.nanoTime() - start
        val durationSeconds = duration / 1e9
        
        return BenchmarkResult(
            test = "String Concatenation (50k iterations)",
            durationNs = duration,
            memoryBytes = 0,
            operations = iterations.toLong(),
            opsPerSec = iterations / durationSeconds
        )
    }
}